tests/end2end/features/keyinput.feature

```diff
@@ -162,3 +162,44 @@ Feature: Keyboard input
         And I press the key "c"
         Then the message "foo 3" should be shown
         And the message "foo 3" should be shown
+
+    # :test all tabs.mode_on_change modes
+
+    Scenario: mode on change normal
+        Given I set tabs.mode_on_change to normal
+        And I clean up open tabs
+        When I open about:blank
+        And I run :enter-mode insert
+        And I open about:blank in a new tab
+        Then "Entering mode KeyMode.insert (reason: command)" should be logged
+        And "Leaving mode KeyMode.insert (reason: tab changed)" should be logged
+        And "normal: mode before tab change: insert." should be logged
+        And "normal: mode after tab change: normal." should be logged
+
+    Scenario: mode on change persist
+        Given I set tabs.mode_on_change to persist
+        And I clean up open tabs
+        When I open about:blank
+        And I run :enter-mode insert
+        And I open about:blank in a new tab
```

Comment:
Loading a new page in the foreground leaves insert mode, which breaks this test (and might make others flaky). You probably want to open the page in the background tab and switch tabs instead.

Reply:
Not sure what you mean here. Do you mean for mode_on_change=normal or persist? Persist mode should not be leaving insert mode.